uwblueprint/project-read-backend
# Implement backend auth

### Notion ticket link



[Set up DRF user authentication](https://www.notion.so/uwblueprintexecs/Set-up-DRF-user-authentication-2ddbd0afa9174e10a864f4a181b05bb5)



### Implementation description
- New secrets are added to the secret vault. 
     - **Re-run `vault kv get -format=json kv/project-read | python ./scripts/update_secret_files.py` to fetch the new secrets**
- Add firebase auth SDK
     - **Re-build containers with `docker-compose build` to install new dependencies**
- Adds an authentication class that expects a token to be passed in the Authorization header, and parses it to get the user and make sure the request is valid. 



### Steps to test

1. Create a user in firebase [here](https://console.firebase.google.com/u/1/project/bp-project-read/authentication/users)
2. Create a user on local, the `uid` field of the local user must be the `uid` of the firebase user
2. Generate a JSON web token with: `curl -d email=FIREBASE_USER_EMAIL -d password=FIREBASE_USER_PASSWORD -d returnSecureToken=true https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword\?key\=AIzaSyCPFRevxMXiQXdWexmJ4fDL48x6HZDJoxE`
3. Copy the JWT and try some requests in Postman (note the authorization type in the screenshot below).
4. It should error out if the JWT is expired (an hour after generation), the `uid` doesn't match any users in your database, or the JWT is invalid. Otherwise, the request should succeed.

![image](https://user-images.githubusercontent.com/22237453/110895491-de9f0b80-82c7-11eb-91d5-1cdf05c6524b.png)



### What should reviewers focus on?

- `authentication.py` and `exceptions.py` in the accounts app
